File: src/main/java/com/iodigital/weather/WeatherApplication.java
package com.iodigital.weather;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.data.r2dbc.repository.config.EnableR2dbcRepositories;
import org.springframework.http.client.reactive.ReactorClientHttpConnector;
import org.springframework.http.client.reactive.ReactorResourceFactory;
import org.springframework.web.reactive.function.client.WebClient;

import java.time.Duration;

@EnableR2dbcRepositories
@SpringBootApplication
public class WeatherApplication {
    @Bean
    public ReactorResourceFactory resourceFactory() {
        return new ReactorResourceFactory();
    }

    @Bean
    public WebClient webClient() {
        return WebClient
            .builder()
            .clientConnector(
                new ReactorClientHttpConnector(
                    resourceFactory(),
                    client -> client.responseTimeout(Duration.ofSeconds(10))
                )
            )
            .build();
    }

    public static void main(String[] args) {
        SpringApplication.run(WeatherApplication.class, args);
    }
}

File: src/main/kotlin/com/iodigital/weather/WeatherApplication.kt
package com.iodigital.weather

import org.springframework.boot.runApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.context.annotation.Bean
import org.springframework.http.client.reactive.ReactorClientHttpConnector
import org.springframework.http.client.reactive.ReactorResourceFactory
import org.springframework.web.reactive.function.client.WebClient
import reactor.blockhound.BlockHound
import reactor.blockhound.integration.BlockHoundIntegration
import java.time.Duration

@SpringBootApplication
class WeatherApplication {
    init {
        BlockHound.install(
            BlockHoundIntegration {
                it
                    .allowBlockingCallsInside(
                        "java.util.ServiceLoader\$LazyClassPathLookupIterator",
                        "parse"
                    )
                    .allowBlockingCallsInside(
                        "kotlin.reflect.jvm.ReflectJvmMapping",
                        "getKotlinFunction"
                    ).allowBlockingCallsInside(
                        "kotlin.reflect.full.KClasses",
                        "getMemberProperties"
                    )
            }
        )
    }

    @Bean
    fun resourceFactory() = ReactorResourceFactory()

    @Bean
    fun webClient() =
        WebClient
            .builder()
            .clientConnector(
                ReactorClientHttpConnector(resourceFactory()) {
                    it.responseTimeout(Duration.ofSeconds(10))
                }
            )
            .build()
}

fun main(args: Array<String>) {
    runApplication<WeatherApplication>(*args)
}
